src/EventInterface.php

```diff
@@ -51,6 +57,7 @@ public function setName($name);
      * Set the event target/context
      *
      * @param  null|string|object $target
+     * @phpstan-param TTarget $target
```

Comment:
We could do `@psalm-self-out` here, to allow the type of the event to change based on `setTarget()` semantics

Reply:
@Ocramius Like so: 4e59ed3691c04b9216828f76a00d73a567c272be? I've implemented it for `setTarget()`, `setParams()` and `__construct()` in `EventInterface` and `Event` to get a feel for it. Never worked with Psalm before, and there's a lot of warnings to address based on that initial setup. I believe this was kinda what you were thinking of right?

Reply:
Something like that, yep - that should help with the statefulness of this event. Can't review now tho: brain is fried :D

Reply:
No rush!